File: LeetCode/Python Solutions/Binary Search Tree to Greater Sum Tree/binary.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def __init__(self):
        self.sum = 0
    
    def bstToGst(self, root: TreeNode) -> TreeNode:
        self.traverse(root)
        return root
    
    def traverse(self, node: TreeNode):
        if not node:
            return
        # Traverse the right subtree first
        self.traverse(node.right)
        
        # Process the current node
        self.sum += node.val
        node.val = self.sum
        
        # Traverse the left subtree
        self.traverse(node.left)

# Example usage:
# sol = Solution()
# root = TreeNode(4)
# root.left = TreeNode(1)
# root.right = TreeNode(6)
# root.left.left = TreeNode(0)
# root.left.right = TreeNode(2)
# root.left.right.right = TreeNode(3)
# root.right.left = TreeNode(5)
# root.right.right = TreeNode(7)
# root.right.right.right = TreeNode(8)

# result = sol.bstToGst(root)
# You can implement a print method to check the results

File: LeetCode/Java Solutions/Binary Search Tree to Greater Sum Tree/binary.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int sum = 0;

    public TreeNode bstToGst(TreeNode root) {
        // Perform a reverse in-order traversal
        traverse(root);
        return root;
    }

    private void traverse(TreeNode node) {
        if (node == null) {
            return;
        }
        // Traverse the right subtree first
        traverse(node.right);

        // Process the current node
        sum += node.val;
        node.val = sum;

        // Traverse the left subtree
        traverse(node.left);
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        // Example usage:
        TreeNode root = new TreeNode(4);
        root.left = new TreeNode(1);
        root.right = new TreeNode(6);
        root.left.left = new TreeNode(0);
        root.left.right = new TreeNode(2);
        root.left.right.right = new TreeNode(3);
        root.right.left = new TreeNode(5);
        root.right.right = new TreeNode(7);
        root.right.right.right = new TreeNode(8);

        TreeNode result = sol.bstToGst(root);

        // Print the resulting GST (you can implement a print method to check the results)
    }
}
